src/experimental/slam_toolbox_lifelong.cpp

```diff
@@ -76,8 +81,374 @@ LifelongSlamToolbox::LifelongSlamToolbox(rclcpp::NodeOptions options)
+  
+  int number_cells = map_dist / resolution;
+  std::vector<std::vector<int>> grid(number_cells); // 20 meters in Y dimension
+  for (int i=0; i<number_cells; ++i)
+  {
+    grid[i].resize(number_cells); // 20 meters in X dimension
+    for (int j=0; j<number_cells; ++j)
+    {
+      grid[i][j] = 0;
+    }
+  }
+
+  // I will have 5 lasers for each reading (Located at 0, +-25, +-50)
+  std::vector<float> robot_pose{5.5f, 6.0f, -PI/2};
+  // std::vector<float> robot_pose{5.5f, 6.0f, PI/2};
+  // std::vector<float> robot_pose{5.5f, 6.0f, 0.0f};
+
+  // This is the initial point
+  std::vector<int> robot_grid_pos = getGridPosition(robot_pose[0], robot_pose[1], resolution);
+  std::cout << "Robot position: " << robot_grid_pos[0] << ", " << robot_grid_pos[1] << std::endl;
+
+  // Creating the laser scan with 5 beams ------- Angles will be -50, -25, 0, 25, 50 //----// +-50 = +-0.87266 : +-25 = +-0.43633
+  std::vector<float> ranges{2.75f, 5.0f, 5.0f, 5.0f, 5.0f}; // Maximum sensor range is 5 meters
+  std::vector<float> angles{0.0f, -0.43633f, 0.0f, 0.43633f, 0.87266f};
+  // std::vector<float> angles{0.785398f, -0.43633f, 0.0f, 0.43633f, 0.87266f};
+  // std::vector<float> angles{0.87266f, -0.43633f, 0.0f, 0.43633f, 0.87266f};
+
+  // Current yaw + beam angle: -PI/2 (-1.570795) -0.87266 = 2.44345 (-55 degrees)
+  // for (int i = 0; i < ranges.size(); ++i)
+  for (int i = 0; i < 1; ++i) // One reading only
+  {
+    std::cout << "........ New laser ........" << std::endl;
+    std::cout << "Distance: " << ranges[i] << ", Angle: " << angles[i] << std::endl;
+
+    // Laser continuous distance
+    std::vector<float> laser_grid = getLaserHit(robot_pose, ranges[i], angles[i]);
+    // Laser final cell
+    std::vector<int> final_grid_pos = getGridPosition(laser_grid[0], laser_grid[1], resolution);
+    // robot_grid_pos[0] // X1 - robot_grid_pos[1] // Y1
+    // final_grid_pos[0] // X2 - final_grid_pos[1] // Y2
+
+    std::vector<int> cells_x, cells_y;
+    std::pair<std::vector<int>, std::vector<int>> res_pair = Bresenham(robot_grid_pos[0], robot_grid_pos[1], final_grid_pos[0], final_grid_pos[1]);
+
+    // Cells visited by this laser beam
+    cells_x = res_pair.first;
+    cells_y = res_pair.second;
+
+    // Adding last hit cell to the set 
+    cells_x.push_back(final_grid_pos[0]); // Here make the change
+    cells_y.push_back(final_grid_pos[1]);
+
+    // Laser beam discretization
+    int samples = 200;
+    float step = ranges[i] / samples;
+    float current_dis = 0.0f;
+
+    // Vectors for discretization of the laser beam
+    std::vector<float> x_coord, y_coord;
+    x_coord.reserve(samples + 1);
+    y_coord.reserve(samples + 1);
+
+    while(current_dis < (ranges[i] + step))
+    {
+      // Discretization loop
+      float x_p = current_dis * sin(angles[i]);
+      float y_p = current_dis * cos(angles[i]);
+      x_coord.emplace_back(x_p);
+      y_coord.emplace_back(y_p);
+      current_dis += step;
+    }
```

Comment:
@kmilo7204 meta: so this is the sampling approach you were referring too. I see what you meant by being somewhat expensive to compute. 

Did you consider using line-to-line intersections? Given a sensor reading, compute the intersections between each beam and each of the sides of a given cell. Discard those intersections that do not fall on the perimeter of the cell. Then, you'll be left with either 2 points or none (the 2 points may be the same point if you the beam hits one of the cell's corners -- I'd disregard that one). Compute distances and use the beam that yields the shortest ones. 

You can optimize it by reversing the search. For a given cell, draw the line that passes through its center and a given sensor reading location. Find the beam that more closely aligns with it. Compute distances for that beam and its neighbors, to left and right. As soon as distances start increasing in a given direction, stop iterating in that direction. When iteration stops, pick the beam that yields the shortest distances. 

We can use Bresenham to narrow down the set of cells, not sure what's the quickest.

Does that make sense?

Reply:
It makes sense.

To be honest I tried so hard to make it work but I am not quite sure about how fast and reliable it could be.

I am testing the line-line intersection. Actually I found an interesting explanation [here](https://www.topcoder.com/thrive/articles/Geometry%20Concepts%20part%202:%20%20Line%20Intersection%20and%20its%20Applications)